pkg/controller/v1alpha1/localmodel/controller.go

```diff
@@ -0,0 +1,658 @@
+		c.Log.Error(err, "unable to list PVCs")
+		return err
+	}
+	for _, pvc := range pvcs.Items {
+		if _, ok := namespaces[pvc.Namespace]; !ok {
+			if pvc.Namespace == jobNamespace {
+				// Keep PVCs in modelCacheNamespace as they don't have a corresponding inference service
+				continue
+			}
+			c.Log.Info("deleting pvc ", "name", pvc.Name, "namespace", pvc.Namespace)
+			persistentVolumeClaims := c.Clientset.CoreV1().PersistentVolumeClaims(pvc.Namespace)
+			if err := persistentVolumeClaims.Delete(context.TODO(), pvc.Name, metav1.DeleteOptions{}); err != nil {
+				c.Log.Error(err, "deleting PVC ", "name", pvc.Name, "namespace", pvc.Namespace)
+			}
+			c.Log.Info("deleting pv", "name", pvc.Name+"-"+pvc.Namespace)
+			persistentVolumes := c.Clientset.CoreV1().PersistentVolumes()
+			if err := persistentVolumes.Delete(context.TODO(), pvc.Name+"-"+pvc.Namespace, metav1.DeleteOptions{}); err != nil {
+				c.Log.Error(err, "deleting PV err")
+			}
+		}
+	}
+
+	for namespace := range namespaces {
+		pv := v1.PersistentVolume{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: localModel.Name + "-" + namespace,
+			},
+			Spec: nodeGroup.Spec.PersistentVolumeSpec,
+		}
+		if err := c.createPV(pv, localModel); err != nil {
+			c.Log.Error(err, "Create PV err", "name", pv.Name)
+		}
+
+		pvc := v1.PersistentVolumeClaim{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      localModel.Name,
+				Namespace: namespace,
+			},
+			Spec: nodeGroup.Spec.PersistentVolumeClaimSpec,
+		}
+		pvc.Spec.VolumeName = pv.Name
+		if err := c.createPVC(pvc, namespace, localModel); err != nil {
+			c.Log.Error(err, "Create PVC err", "name", pvc.Name)
+		}
+	}
+	return nil
+}
+
+// Step 1 - Checks if the CR is in the deletion process, if so, it creates deletion jobs to delete models on all nodes.
+// Step 2 - Creates PV & PVC for model download
+// Step 3 - Creates Jobs on all nodes to download models
+// Step 4 - Creates PV & PVCs for namespaces with isvcs using this cached model
+func (c *LocalModelReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	c.Log.Info("Reconcile() called for localmodel", "name", req.Name)
+
+	localModelConfig, err := v1beta1.NewLocalModelConfig(c.Clientset)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	defaultImage = localModelConfig.DefaultImage
+
+	localModel := &v1alpha1api.ClusterLocalModel{}
+	if err := c.Get(ctx, req.NamespacedName, localModel); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	nodeGroup := &v1alpha1api.LocalModelNodeGroup{}
+	nodeGroupNamespacedName := types.NamespacedName{Name: localModel.Spec.NodeGroup}
+	if err := c.Get(ctx, nodeGroupNamespacedName, nodeGroup); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Step 1 - Checks if the CR is in the deletion process, if so, creates deletion jobs to delete models on all nodes.
+	if localModel.ObjectMeta.DeletionTimestamp.IsZero() {
```

Comment:
Should here be checking `DeleteTimestamp` is not null ?

Reply:
No, this is checking not deleted -> add finalizer.